### PR TITLE
feat: add support for nested file paths in latch metadata

### PR DESCRIPTION
**Pull request by rahuldesai1**
The purpose of this is to support nested file paths in the `generate-metadata` command. 

The interface is a bit more complicated; however, this allows users to be able to upload Snakemake workflows without having to make any modifications to their config.yaml (which can sometimes be quite large). 

Not planning on merging until docs are updated.

### CODE REVIEW

File: latch/types/metadata.py [hunk 1]
```diff
@@ -393,6 +395,7 @@ class SnakemakeParameter(LatchParameter):
     default: Optional[Any] = None
 
 
+# DEPRECATED: use `file_metadata` keyword in `SnakemakeMetadata` instead
```
Comment (maximsmol) [inline latch/types/metadata.py:398]: find some way to do https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated

at the very least put this in the docstring

File: latch/types/metadata.py [hunk 2]
```diff
@@ -418,6 +421,22 @@ class SnakemakeFileParameter(SnakemakeParameter):
     """
 
 
+@dataclass
+class SnakemakeFileMetadata:
+    path: Optional[Path] = None
+    """
+    The path where the file passed to this parameter will be copied.
```
Comment (maximsmol) [inline latch/types/metadata.py:428]:
```suggestion
    The path where the file passed to this parameter will be copied
```

File: latch/types/metadata.py [hunk 3]
```diff
@@ -418,6 +421,22 @@ class SnakemakeFileParameter(SnakemakeParameter):
...
+    """
+    config: bool = False
+    """
+    Whether or not the file path is exposed in the Snakemake config
```
Comment (maximsmol) [inline latch/types/metadata.py:432]:
```suggestion
    If `True`, expose the file in the Snakemake config
```
1. "If `True`" phrasing is from Google docstring style guide
2. active voice

File: latch/types/metadata.py [hunk 4]
```diff
@@ -418,6 +421,22 @@ class SnakemakeFileParameter(SnakemakeParameter):
...
+    """
+    download: bool = False
+    """
+    Whether or not the file is downloaded in the JIT step
```
Comment (maximsmol) [inline latch/types/metadata.py:436]:
```suggestion
    If `True`, download the file in the JIT step
```

File: latch_cli/snakemake/workflow.py [hunk 5]
```diff
@@ -301,9 +295,24 @@ def interface_to_parameters(
...
-                ctx, default, python_type=interface.inputs[k], expected=v.type
-            )
+            try:
+                default_lv = TypeEngine.to_literal(
+                    ctx, default, python_type=interface.inputs[k], expected=v.type
+                )
+            except TypeTransformerFailedError as e:
+                if "Python value cannot be None" in str(e):
+                    click.secho(
+                        reindent(
```
Comment (maximsmol) [inline latch_cli/snakemake/workflow.py:305]: `dedent`

File: latch_cli/snakemake/workflow.py [hunk 6]
```diff
@@ -387,86 +397,193 @@ def get_fn_return_stmt(self):
...
+
             """,
-            1,
+            0,
         )
 
-        for param, t in self.python_interface.inputs.items():
-            param_meta = self.parameter_metadata[param]
+        if config:
+            code_block += reindent(
```
Comment (maximsmol) [inline latch_cli/snakemake/workflow.py:457]: `dedent`

File: latch_cli/snakemake/workflow.py [hunk 7]
```diff
@@ -387,86 +397,193 @@ def get_fn_return_stmt(self):
...
+                param_meta.path,
+                param_meta.download,
+                param_meta.config,
+                blob_params,
+            )
+
+            return code_block
+
+        if file_meta is None or file_meta.get(param) is None or is_primitive_type(t):
+            return reindent(
```
Comment (maximsmol) [inline latch_cli/snakemake/workflow.py:497]: `dedent`

File: latch_cli/snakemake/workflow.py [hunk 8]
```diff
@@ -387,86 +397,193 @@ def get_fn_return_stmt(self):
...
+                meta.path,
+                meta.download,
+                meta.config,
+                blob_params,
+            )
 
+        code_blocks: List[str] = []
+        if is_list_type(t):
+            code_blocks.append(
+                reindent(
```
Comment (maximsmol) [inline latch_cli/snakemake/workflow.py:520]: `dedent`

**Review by maximsmol — CHANGES_REQUESTED** [review]

File: latch_cli/snakemake/workflow.py [hunk 9]
```diff
@@ -387,86 +397,170 @@ def get_fn_return_stmt(self):
...
+            )
+
+            for i, m in enumerate(meta):
+                sub_param = f"{param}[{i}]"
+                code_blocks.append(
+                    self.get_param_code(
+                        t.__args__[0],
+                        sub_param,
+                        {sub_param: m},
+                        blob_params + [i],
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:520]:
```suggestion
                        [*blob_params, i],
```

stylistic nit

File: latch_cli/snakemake/workflow.py [hunk 10]
```diff
@@ -387,86 +397,170 @@ def get_fn_return_stmt(self):
...
+                        sub_param,
+                        {sub_param: meta.get(field.name)},
+                        blob_params + [field.name],
+                    )
                 )
 
-            if not getattr(param_meta, "config", True):
-                continue
+        return dedent(f"""\
+            __code__\n""").replace("__code__", "\n".join(code_blocks))
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:540]:
```suggestion
        return "\n".join(code_blocks)
```
?

**Review by ayushkamat — CHANGES_REQUESTED** [review]

File: latch_cli/snakemake/config/utils.py [hunk 11]
```diff
@@ -115,8 +118,23 @@ def parse_type(
...
+        if typ in {LatchFile, LatchDir}:
+            click.secho(
+                dedent(f"""
+                Failed to parse {name}. Lists containing LatchFile or LatchDir
+                are not supported.
+                """),
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+            # TODO: (rahul) add back once console PR is merged
```
Comment (ayushkamat) [inline latch_cli/snakemake/config/utils.py:135]:
```suggestion
            # todo(rahul): add back once console PR is merged
```

minor stylistic thing we do

File: latch_cli/snakemake/config/utils.py [hunk 12]
```diff
@@ -148,11 +166,14 @@ def parse_value(t: Type, v: JSONValue):
         return v, v
 
     if isinstance(v, list):
-        assert get_origin(t) is list
+        assert get_origin(t) is list or get_origin(t) is Annotated
```
Comment (ayushkamat) [inline latch_cli/snakemake/config/utils.py:169]: would actually put this at the top

```
def parse_value(t: Type, v: JSONValue):
    if get_origin(t) is Annotated:
        args = get_args(t)
        assert len(args) > 0
        
        return parse_value(args[0], v)
```

this way we support any valid annotated type

File: latch_cli/snakemake/config/utils.py [hunk 13]
```diff
@@ -148,11 +166,14 @@ def parse_value(t: Type, v: JSONValue):
...
+        if get_origin(t) is Annotated:
+            t = get_args(t)[0]
 
         sub_type = get_args(t)[0]
         res = [parse_value(sub_type, x) for x in v]
-        return [x[0] for x in res], [x[1] for x in res]
+        val, defaults = [x[0] for x in res], [x[1] for x in res]
+        return val, None if any([x is None for x in defaults]) else defaults
```
Comment (ayushkamat) [inline latch_cli/snakemake/config/utils.py:176]: weird, why this change?

File: latch_cli/snakemake/config/utils.py [hunk 14]
```diff
@@ -161,14 +182,18 @@ def parse_value(t: Type, v: JSONValue):
...
 
+    ignore_default = False
     for k, x in v.items():
         sanitized = identifier_from_str(k)
         assert sanitized in fs, sanitized
         val, default = parse_value(fs[sanitized].type, x)
         ret[sanitized] = val
+        # flytekit fails if any fields in default object are None
+        if default is None:
+            ignore_default = True
```
Comment (ayushkamat) [inline latch_cli/snakemake/config/utils.py:193]: can probably del

File: latch_cli/snakemake/workflow.py [hunk 15]
```diff
@@ -496,34 +482,41 @@ def get_param_code(
...
-            code_blocks.append(dedent(f"""
-                    if len({param}) != {len(meta)}:
-                        raise ValueError("The size of input files list must be equal to the number of `SnakemakeFileMetadata` objects provided in latch_metadata")
-                    """))
+            sub_typ = get_args(t)[0]
+            if sub_typ in (LatchFile, LatchDir):
+                click.secho(
+                    dedent(f"""
+                    Failed to parse {param}. Lists containing LatchFile or LatchDir
+                    are not supported.
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:495]:
```suggestion
                    are not yet supported.
```

File: latch_cli/snakemake/workflow.py [hunk 16]
```diff
@@ -496,34 +482,41 @@ def get_param_code(
...
+                param_path,
             )
 
         code_blocks: List[str] = []
         if is_list_type(t):
-            code_blocks.append(dedent(f"""
-                    if len({param}) != {len(meta)}:
-                        raise ValueError("The size of input files list must be equal to the number of `SnakemakeFileMetadata` objects provided in latch_metadata")
-                    """))
+            sub_typ = get_args(t)[0]
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:490]: idr if you guard against generic lists above this but if you don't may wanna put that here

**Review by ayushkamat — CHANGES_REQUESTED** [review]

File: latch/types/metadata.py [hunk 17]
```diff
@@ -602,8 +615,49 @@ class SnakemakeMetadata(LatchMetadata):
...
+                for val in param:
+                    if _contains_none_fields(list_typ, val):
+                        return True
+                return False
+
+            assert is_dataclass(t)
+            for field in fields(t):
+                if _contains_none_fields(field.type, getattr(param, field.name)):
+                    return True
+            return False
```
[the remark this answers is not in the document]
Comment (rahuldesai1) [reply latch/types/metadata.py:644]: TODO: per our conversation, instead of checking null values, do type checking to verify that default parameters match the expected type so that flyte doesn't blow up.

File: latch_cli/snakemake/config/utils.py (hunk 13 above)
Comment (rahuldesai1) [reply latch_cli/snakemake/config/utils.py:176]: Since we don't provide defaults for LatchFile/LatchDir, the default would be set to None, which would cause Flyte to error. 

It seems like another option is to set the field to None and require that the user manually set the default.

File: latch/types/metadata.py [hunk 18]
```diff
@@ -439,7 +427,7 @@ class SnakemakeFileParameter(SnakemakeParameter):
 
 @dataclass
 class SnakemakeFileMetadata:
-    path: Optional[Path] = None
+    path: Path = None
```
Comment (ayushkamat) [inline latch/types/metadata.py:430]: don't give it a default

File: latch/types/metadata.py [hunk 19]
```diff
@@ -621,39 +609,11 @@ class SnakemakeMetadata(LatchMetadata):
...
                 continue
-            if _contains_none_fields(param.type, param.default):
-                click.secho(
-                    " Non-None default parameters cannot contain None fields. Please"
-                    " remove default or set all fields to a non-None value for"
-                    f" parameter: {name}.  ",
-                    fg="red",
-                )
-                raise click.exceptions.Exit(1)
+            validate_snakemake_type(name, param.type, param.default)
```
Comment (ayushkamat) [inline latch/types/metadata.py:616]: try-except this, `click.secho` the caught `ValueError`'s message, and then raise `click.exceptions.Exit(1)`

**Review by ayushkamat — COMMENTED** [review]
kinda strange that the type validation we're doing here doesn't support general unions and handles optionals as an edge case tbh, but its not a big deal for now
